lightning/src/ln/msgs.rs

```diff
@@ -887,6 +888,39 @@ mod fuzzy_internal_msgs {
 		// 12 bytes of 0-padding for Legacy format
 	}
 
+	pub(crate) trait OnionHopDataLength: Writeable {
+		fn encoding_length(&self) -> usize;
+	}
+
+	impl OnionHopDataLength for OnionHopData {
+		fn encoding_length(&self) -> usize {
+			fn high_zero_bytes_dropped_length(data: u64) -> usize {
+				8 - ((data.leading_zeros()/8) as usize)
+			}
+			match self.format {
+				OnionHopDataFormat::Legacy { .. } => 1 + 8 + 8 + 4 + 12, // 33
+				OnionHopDataFormat::NonFinalNode { .. } => {
+					1 +
+						1 + 1 + high_zero_bytes_dropped_length(self.amt_to_forward) +
+						1 + 1 + high_zero_bytes_dropped_length(self.outgoing_cltv_value as u64) +
+						1 + 1 + 8 // short_channel_id
+				}
+				OnionHopDataFormat::FinalNode { ref payment_data } => {
+					1 +
+						1 + 1 + high_zero_bytes_dropped_length(self.amt_to_forward) +
+						1 + 1 + high_zero_bytes_dropped_length(self.outgoing_cltv_value as u64) +
+						match payment_data {
+							None => 0,
+							Some(payment_data) =>
+								1 + 1 + high_zero_bytes_dropped_length(payment_data.total_msat) +
+									32 // payment_secret
```

Comment:
Well, `payment_data` would need an `unwrap` in the else branch, which I usually try to avoid if I need to do something in both some/none cases.  Just aesthetics, so happy to change it if LDK is standardizing on the if/else form.